### ISSUE TITLE
Creating a theme at startup and enabling it

### ISSUE THREAD

**Issue by qcoumes** [fabiocaccamo/django-admin-interface#99]
**Python version**
3.8

**Django version**
3.1

**Package version**
0.14.0

**Current behavior (bug description)**
I am creating a theme from `settings` in an `AppConfig.ready()`, the Theme models is correctly created with the right values. `active` is set to `True` (and is the only Theme ticked in the admin panel), but the Django's default one is the theme applied. I have to manually save the model once from the Django admin panel for it to be applied.

Here's my `ready()` function creating the Theme at startup :

```python
class WebsiteConfig(AppConfig):
    name = 'website'
    
    
    def ready(self):
        if not {"runserver", "shell"}.intersection(sys.argv):
            return
        
        Theme = apps.get_model("admin_interface", "Theme")  # noqa
        kwargs = {
            "name":                                     "ISLRN",
            "active":                                   True,
            "title":                                    "ISLRN Administration",
            "title_color":                              settings.COLOR_PALETTE["theme_on_primary"],
            "title_visible":                            True,
            "logo":                                     "",
            "logo_color":                               "#FFFFFF",
            "logo_visible":                             False,
            "css_header_background_color":              settings.COLOR_PALETTE["theme_primary"],
            "css_header_text_color":                    settings.COLOR_PALETTE["theme_on_primary"],
            "css_header_link_color":                    settings.COLOR_PALETTE["theme_on_primary_lighter"],
            "css_header_link_hover_color":              settings.COLOR_PALETTE["theme_on_primary_darker"],
            "css_module_background_color":              settings.COLOR_PALETTE["theme_secondary"],
            "css_module_text_color":                    settings.COLOR_PALETTE["theme_on_secondary"],
            "css_module_link_color":                    settings.COLOR_PALETTE["theme_on_secondary_lighter"],
            "css_module_link_hover_color":              settings.COLOR_PALETTE["theme_on_secondary_darker"],
            "css_module_rounded_corners":               True,
            "css_generic_link_color":                   settings.COLOR_PALETTE["theme_on_default_lighter"],
            "css_generic_link_hover_color":             settings.COLOR_PALETTE["theme_on_default_darker"],
            "css_save_button_background_color":         settings.COLOR_PALETTE["theme_success"],
            "css_save_button_background_hover_color":   settings.COLOR_PALETTE["theme_success"],
            "css_save_button_text_color":               settings.COLOR_PALETTE["theme_on_success"],
            "css_delete_button_background_color":       settings.COLOR_PALETTE["theme_error"],
            "css_delete_button_background_hover_color": settings.COLOR_PALETTE["theme_error"],
            "css_delete_button_text_color":             settings.COLOR_PALETTE["theme_on_error"],
            "css":                                      "",
            "related_modal_active":                     True,
            "related_modal_background_color":           settings.COLOR_PALETTE["theme_default"],
            "related_modal_background_opacity":         0.2,
            "related_modal_rounded_corners":            True,
            "list_filter_dropdown":                     False,
            "recent_actions_visible":                   True
        }
        
        try:
            theme = Theme.objects.get(name="ISLRN")
            for k, a in kwargs.items():
                setattr(theme, k, a)
            theme.save()
        except Theme.DoesNotExist:
            Theme.objects.create(**kwargs)
        Theme.get_active_theme()
```

**Expected behavior**
I expected the created Theme to be applied immediately since it's the only Theme with `active` set to `True`.

**Comment by fabiocaccamo** [comment]
@qcoumes the purpose of this application is to manage/customize themes using the admin.

As you can see [here](https://github.com/fabiocaccamo/django-admin-interface/blob/master/admin_interface/apps.py), some stuff is done on app ready, I think that your usage have some conflicts with it. 

Your application is listed before or after `admin_interface` in `settings.INSTALLED_APPS`?

**Comment by qcoumes** [comment]
@fabiocaccamo Thanks for your help

> @qcoumes the purpose of this application is to manage/customize themes using the admin.

I figured it would be easier to use this application instead of modifying the django's template files.


> As you can see [here](https://github.com/fabiocaccamo/django-admin-interface/blob/master/admin_interface/apps.py), some stuff is done on app ready, I think that your usage have some conflicts with it.
> 
> Your application is listed before or after `admin_interface` in `settings.INSTALLED_APPS`?

It was listed before. I tried to list `admin_interfaces` first, but the same problem occurs.

I can upload a minimal example if you want.

**Comment by merwok** [comment]
Just curious: is there a reason for your use of try: create / except: get vs. Theme.objects.get_or_create?

@fabiocaccamo I suppose the admin or admin form for Theme does something to ensure only one theme at a time is active, right?  So creating a new Theme with `active=True` is not enough!  (unless the something is done in Theme.save, which wouldn’t be a clean django practice — model methods should work with one row, manager or queryset methods with multiple rows).

**Comment by qcoumes** [comment]
> Just curious: is there a reason for your use of try: create / except: get vs. Theme.objects.get_or_create?

@merwok  Mainly because I don't know which field is required for a new instance (I guess `name` is not enough). Also to avoid the loop over `kwargs` and the use of `setattr` if it is not needed.

**Comment by fabiocaccamo** [comment]
@merwok the active theme is chosen [here](https://github.com/fabiocaccamo/django-admin-interface/blob/master/admin_interface/models.py#L53) and this method is called by different signals.

@qcoumes if you want write a PR covered by the relative test case I would be glad to merge it.

**Comment by merwok** [comment]
@qcoumes I didn’t understand your reply!  You say you want to avoid loop over kwargs + setattr, but the code you pasted has that, and using get_or_create would avoid it, so it’s confusing :slightly_smiling_face: 

@fabiocaccamo ah I see the get_active method is using `Theme.objects.filter(active=True).first()`, but in the absence of `order_by` method or `Meta.ordering` property on model it’s not explicit which object would be returned (I don’t rely on implicit sorts so I don’t remember if django would order by ID descending or ascending, or if that depends on the database server).

**Comment by qcoumes** [comment]
@merwok I only loop over `kwargs` when modifying the `Theme` if it already exists, not when creating it. Won't `get_or_create` create a new instance every time one of the settings are modified?

@fabiocaccamo I can't find where the issue come from, I tried to manually set the cached theme by calling `set_cached_active_theme()`, a `print()` inside `get_cached_active_theme()` correctly displays my theme, yet the django administration is still displaying Django's default theme.

Note that by Django's theme, I mean the blue vanilla one, not the green one created by `django-admin-interface`.

**Comment by merwok** [comment]
Ah I misread the first time!  Indeed, you need [`update_or_create`](https://docs.djangoproject.com/en/3.1/ref/models/querysets/#update-or-create) for your case.

**Comment by qcoumes** [comment]
@merwok Didn't know such method existed, this is indeed exactly what I need, thanks.

**Comment by fabiocaccamo** [comment]
@qcoumes I think that the problem is that your theme is created before the default one, could you please check and tell me the ids of the two themes?

**Comment by qcoumes** [comment]
Sorry for the (very) late reply.

The default one is created before my theme. `Django`'s pk is `1` and `ISLRN` pk is 2.

**Comment by qcoumes** [comment]
I actually found the problem while creating a minimum reproducible example.

Each color of the setting `settings.COLOR_PALETTE` only contains the hexcode, but without the starting `#`, so none of the colour field was valid. So my theme was correctly selected, but I guess it used the default theme color for these invalid field.

Then when I opened the admin's form, HTML5 colour field were smart enough to add the missing `#`, hence why it worked after saving the theme once.
